File: scripts/algorithms/F/Find in Mountain Array/Find in Mountain Array.cpp
// Runtime: 13 ms (Top 8.65%) | Memory: 6.9 MB (Top 88.34%)
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 * public:
 * int get(int index);
 * int length();
 * };
 */

class Solution {
public:

    //----------------------- find peak ---------------------------------------

    int find_max(MountainArray &mountainArr,int start,int end)
    { int n= mountainArr.length();
        while(start<=end)
        { int mid= start+(end-start)/2;
           int next=(mid+1)%n;
          int prev=(mid+n-1)%n;

         int mid_val = mountainArr.get(mid);
         int pre_val = mountainArr.get(prev);
         int next_val = mountainArr.get(next);

           if(mid_val > next_val and mid_val>pre_val)
               return mid;

           else if(mid_val<next_val and mid_val > pre_val)
                start=mid+1;

           else if(mid_val>next_val and mid_val<pre_val)
                end=mid-1;

        }
        return -1;

    }

   //--------------------------binary search-------------------------------------------

   int binary_search(MountainArray &mountainArr,int start,int end,int target)
   {
       while(start<=end)
       {
           int mid= start + (end-start)/2;
             int mid_val=mountainArr.get(mid);

           if(mid_val==target)
               return mid;
          else if(target < mid_val)
           {
               end=mid-1;
           }
           else
               start=mid+1;
       }
       return -1;
   }

    //----------------------binary search in reverse sorted------------------------------

       int binary_search_rev(MountainArray &mountainArr,int start,int end,int target)
   {
       while(start<=end)
       {
           int mid= start + (end-start)/2;
           int mid_val=mountainArr.get(mid);
           if(mid_val==target)
               return mid;
          else if(target > mid_val)
           {
               end=mid-1;
           }
           else
               start=mid+1;
       }
       return -1;
   }

  //------------------------------returns minimum index of target--------------------------------------

    int evaluate_ans(int a,int b)
    {
        if(a==-1 and b==-1 )
            return -1;
        else if(a!= -1 and b!= -1)
            return min(a,b);
        else if(a==-1 and b!=-1)
            return b;
        else
            return a;

    }

    int findInMountainArray(int target, MountainArray &mountainArr) {

        int start=0;
        int n= mountainArr.length()-1;
        int max_in = find_max(mountainArr,start ,n);

        int a= binary_search(mountainArr,start,max_in,target);
        int b= binary_search_rev(mountainArr,max_in + 1,n,target);

        return evaluate_ans(a,b);
    }
};

File: scripts/algorithms/F/Find in Mountain Array/Find in Mountain Array.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class Solution:
    def findInMountainArray(self, target: int, mountain_arr: 'MountainArray') -> int:
        min_index = -1
        peak = self.findpeak(mountain_arr)
        
        min_index = self.binary_search(0, peak, mountain_arr, target, 1)
        if min_index == -1:
            min_index = self.binary_search(peak+1, mountain_arr.length() - 1, mountain_arr, target, -1)
        return min_index

    def findpeak(self, mountain_arr):
        start = 0
        end = mountain_arr.length() - 1
        while start < end:
            mid = start + int((end - start)/2)
            if mountain_arr.get(mid) < mountain_arr.get(mid + 1):
                start = mid + 1
            else:
                end = mid
        
        return start
    
    def binary_search(self, start, end, mountain_arr, target, asc):
        while start <= end:
            mid = start + int((end - start)/2)
            mountain_arr_get_mid = mountain_arr.get(mid)
            if target == mountain_arr_get_mid:
                return mid
            if asc == 1:
                if target < mountain_arr_get_mid:
                    end = mid - 1
                elif target > mountain_arr_get_mid:
                    start = mid + 1
            else:
                if target < mountain_arr_get_mid:
                    start = mid + 1
                elif target > mountain_arr_get_mid:
                    end = mid - 1
            
        return -1






